### PR TITLE
feat: add spinning progress UX for `DataCommand` and `MetricsCommand`

### PR DESCRIPTION
**Pull request by griff-rees**
Add `Spinners::Dots` from the [`spinners`](https://lib.rs/crates/spinners) package to improve the UX.

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -86,12 +89,18 @@ impl RunCommand for DataCommand {
...
+        );
         let popgetter = Popgetter::new_with_config(config).await?;
+        let search_results = popgetter.search(self.search_params_args.clone().into());
 
-        let mut data = popgetter
-            .search(self.search_params_args.clone().into())
-            .download(&popgetter.config)
-            .await?;
+        sp.stop();
+
```
Comment (penelopeysm) [inline src/cli.rs:100]: I think feature wise it would be nice to print the number of results here so that the user can mash ctrl-c if they see it trying to download 200,000 metrics while connected to mobile hotspot :)

```suggestion
        let len_requests = search_results.0.shape().0;
        println!("Found {} metrics.", len_requests);
```

Yet another upgrade on this might actually be to print this and prompt for confirmation that the user indeed wants to download this. We could add a `-y` command line option to `DataCommand` which, if passed, will skip the prompt. GitHub doesn't let me comment on lines that are not changed by the PR but the command-line args would be the section which begins with `pub struct DataCommand` in this file.

Comment (griff-rees) [reply src/cli.rs:100]: Ah cool, I've never seen this "Add suggestion to batch" option either. Think I'll stick with the "commit suggestion" and have a go with the command line bit.

File: src/cli.rs [hunk 2]
```diff
@@ -86,12 +89,19 @@ impl RunCommand for DataCommand {
...
+        let search_results = popgetter.search(self.search_params_args.clone().into());
 
-        let mut data = popgetter
-            .search(self.search_params_args.clone().into())
-            .download(&popgetter.config)
-            .await?;
+        sp.stop();
+        let len_requests = search_results.0.shape().0;
+        println!("Found {} metrics.", len_requests);
+        let mut sp = Spinner::new(DEFAULT_PROGRESS_SPINNER, "Download Parquet files".into());
```
Comment (sgreenbury) [inline src/cli.rs:102]: Perhaps reword this to:
```suggestion
        let mut sp = Spinner::new(DEFAULT_PROGRESS_SPINNER, "Downloading metrics".into());
```

File: src/cli.rs [hunk 3]
```diff
@@ -259,8 +269,13 @@ impl RunCommand for MetricsCommand {
         info!("Running `metrics` subcommand");
         debug!("{:#?}", self);
 
+        let mut sp = Spinner::new(
+            DEFAULT_PROGRESS_SPINNER,
+            "Downloading metadata from Azure".into(),
```
Comment (sgreenbury) [inline src/cli.rs:274]: Perhaps reword this one too (since will not necessarily be from Azure)?
```suggestion
            "Downloading and searching metadata".into(),
```

File: src/cli.rs [hunk 4]
```diff
@@ -259,8 +269,13 @@ impl RunCommand for MetricsCommand {
...
+        );
         let popgetter = Popgetter::new_with_config(config).await?;
         let search_results = popgetter.search(self.search_params_args.clone().into());
+        sp.stop();
```
Comment (sgreenbury) [inline src/cli.rs:278]:
```suggestion
        sp.stop_with_newline();
```

**Comment by sgreenbury** [comment]
Thanks for this @griff-rees, I think this looks great and really like having the progress feedback when RUST_LOG is not set - shall we go ahead and merge this into main?

**Comment by griff-rees** [comment]
Thanks @sgreenbury :). I'm trying to work through @penelopeysm's suggestion now, sorry if that's made it a bit more complicated, think I'm most of the way there and it seems compatible with your suggestions which I've merged.

**Comment by griff-rees** [comment]
@penelopeysm: I think this is approaching your suggestion.

I've left some comments around
-  line 112 where the follwoing option is commented out: `sp.stop_and_persist(&COMPLETE_PROGRESS_STRING, spinner_message.into());`. If the timing could be accessed, the formatting could be nicer
- line 260 where `fn print_metrics_count` is defined, which *could* be refactored to manage the logging process in multiple places.

Other thoughts (which may be separate tickets if of interest):
- Try replacing `const DEFAULT_PROGRESS_SPINNER: Spinners = Spinners::Dots;` with `const DEFAULT_PROGRESS_SPINNER: Spinners = Spinners::Earth;`
- Test of the cli so I can automate running these variations... see [this](https://rust-cli.github.io/book/tutorial/testing.html#testing-cli-applications-by-running-them)

File: src/cli.rs [hunk 5]
```diff
@@ -213,6 +254,15 @@ fn parse_year_range(value: &str) -> Result<Vec<YearRange>, &'static str> {
         .collect::<Result<Vec<YearRange>, &'static str>>()
 }
 
+// A simple function to manage similaries across multiple cases.
+// May ultimately be generalised to a function to manage all progress UX
+// that can be switched on and off.
+fn print_metrics_count(search_results: SearchResults) -> usize {
+    let len_requests = search_results.0.shape().0;
+    println!("Found {len_requests} metric(s).");
```
Comment (sgreenbury) [inline src/cli.rs:262]: Potentially this, but could be included in the next PR for other refinements:
```suggestion
    if len_requests == 1 {
        println!("Found {len_requests} metric.");
    } else {
        println!("Found {len_requests} metrics.");
    }
```

File: src/cli.rs [hunk 6]
```diff
@@ -419,4 +473,10 @@ mod tests {
         let output_format = OutputFormat::from_str("awesome_tiny_model");
         assert!(output_format.is_err(), "non listed formats should fail");
     }
+
+    #[test]
+    fn cli() {
+        use clap::CommandFactory;
+        Cli::command().debug_assert();
```
Comment (sgreenbury) [inline src/cli.rs:480]: Helpful to have this test given does not seem to be known until runtime.

**Review by sgreenbury — APPROVED** [review]
Thanks for this @griff-rees and @penelopeysm, I this looks great to merge with your latest changes @griff-rees and we can open another issue/PR for further enhancements.
